Src/Support/GoogleApis.Core/Apis/Util/UriPatcher.cs

```diff
@@ -0,0 +1,112 @@
+    //
+    // Unless IDN/IRI parsing is enabled -- which it is not, by default, prior to .NET 4.5 --
+    // Uri.EscapeDataString uses the set of "unreserved" characters from RFC 2396 instead of the
+    // newer, *smaller* list from RFC 3986. We build requests using URI templating as described
+    // by RFC 6570, which specifies that the latter definition (RFC 3986) should be used.
+    //
+    // This breaks API requests with parameters including any of: !*'()
+    //
+    // == Solutions
+    //
+    // Though the default behaviors changed in .NET 4.5, these "quirks" remain for compatibility
+    // unless the application explicitly targets the new runtime.  Usually, that means adding a
+    // TargetFrameworkAttribute to the entry assembly.
+    //
+    // Applications running on .NET 4.0 or later can also set "DontUnescapePathDotsAndSlashes"
+    // and enable IDN/IRI parsing using app.config or web.config.
+    //
+    // As a class library, we can't control app.config or the entry assembly, so we can't take
+    // either approach. Instead, we resort to reflection trickery to try to solve these problems
+    // if we detect they exist. Sorry.
+    internal class UriPatcher
+    {
+        public static void PatchUriQuirks()
+        {
+            var uriParser = typeof(System.Uri).Assembly.GetType("System.UriParser");
+            if (uriParser == null) { return; }
+
+            // Is "%2f" unescaped for http: or https: URIs?
+            if (new Uri("http://example.com/%2f").AbsolutePath == "//" ||
+                new Uri("https://example.com/%2f").AbsolutePath == "//")
+            {
+                // Call System.UriParser.Http[s]Uri.SetUpdatableFlags(UriSyntaxFlags.None)
+                // https://github.com/Microsoft/referencesource/blob/d925d870f3cb3f6a/System/net/System/_UriSyntax.cs#L87
+                // https://github.com/Microsoft/referencesource/blob/d925d870f3cb3f6a/System/net/System/_UriSyntax.cs#L77
+                // https://github.com/Microsoft/referencesource/blob/d925d870f3cb3f6a/System/net/System/_UriSyntax.cs#L352
+
+                var setUpdatableFlagsMethod = uriParser.GetMethod("SetUpdatableFlags",
+                    BindingFlags.Instance | BindingFlags.NonPublic);
+                if (setUpdatableFlagsMethod != null)
+                {
+                    Action<string> setUriParserUpdatableFlags = (fieldName) =>
+                    {
+                        var parserField = uriParser.GetField(fieldName,
+                            BindingFlags.Static | BindingFlags.NonPublic);
+                        if (parserField == null) { return; }
+                        var parserInstance = parserField.GetValue(null);
+                        if (parserInstance == null) { return; }
+                        setUpdatableFlagsMethod.Invoke(parserInstance, new object[] { 0 });
+                    };
+
+                    // Make the change for the http: and https: URI parsers.
```

Comment:
Yes, this changes global behavior. If someone else changes it too, maybe they have an even better reason than us? :)

We're not prepared for every eventuality. But I'm pretty confident that programs that use us won't have another library that's also futzing with URI quirks mode in a way that would interfere with us.